pkg/statistics/handle/globalstats/topn.go

```diff
@@ -136,9 +264,180 @@ func MergeGlobalStatsTopNByConcurrency(
 //     but is not placed to global-level TopN. We should put them back to histogram latter.
 //  3. `[]*Histogram` are the partition-level histograms which
 //     just delete some values when we merge the global-level topN.
+//
+// The function use the merging operation based on the property that the topn items are ordered.
+// And it does a heuristic short-cutting:
+// - for newly seen merged topn item, it's total occurrence is the sum in the topn and the sum in the hist.
+// - its occurrence in each hist is calculated as row_num_hist / ndv_in_hist or 0 or the bucket bound.
+// - so we can calculate its maximum possible added value to max(row_num_hist / ndv_in_hist, bucket bound of each bucket) for each hist + sum_in_topn
+// - we don't need to actually check it histogram by histogram if the maximum possible occurrence is still smaller than smallest maintained topn.
+// This short-cutting will save the CPU time.
 func MergePartTopN2GlobalTopN(
 	loc *time.Location,
-	version int,
+	topNs []*statistics.TopN,
+	n uint32,
+	hists []*statistics.Histogram,
+	isIndex bool,
+	killer *sqlkiller.SQLKiller,
+) (*statistics.TopN, []statistics.TopNMeta, []*statistics.Histogram, error) {
+	var mergingHeap topnHeap = make([]*heapItem, 0, len(topNs))
+	sumTopN := 0
+	for i, topN := range topNs {
+		if topN.Num() == 0 {
+			continue
+		}
+		sumTopN += topN.Num()
+		heap.Push(&mergingHeap, &heapItem{
+			item:          &topN.TopN[0],
+			idx:           i,
+			nextPosInTopN: 1,
+		})
+	}
+	if mergingHeap.Len() == 0 {
+		return nil, nil, hists, nil
+	}
+	maxPossibleAdded := make([]int64, len(hists))
+	for i, hist := range hists {
+		curMax := int64(hist.NotNullCount() / float64(hist.NDV))
+		for _, bkt := range hist.Buckets {
+			curMax = max(curMax, bkt.Repeat)
+		}
+		maxPossibleAdded[i] = curMax
+	}
+	histIters := make([]histIter, len(hists))
+	for i, hist := range hists {
+		histIters[i].hist = hist
+	}
+	type maintaining struct {
+		affectedTopNs *bitset.BitSet
+		item          statistics.TopNMeta
+	}
+	cur := maintaining{
+		affectedTopNs: bitset.New(uint(len(hists))),
+	}
+	var finalTopNs topNMeataHeap = make([]statistics.TopNMeta, 0, n+1)
+	remainedTopNs := make([]statistics.TopNMeta, 0, n)
+	affectedHist := make([]int, 0, len(hists))
+	firstTime := true
+	checkTheCurAndMoveForward := func(nextVal *statistics.TopNMeta, position uint) error {
+		// It's perf-sensitive path. Don't use defer.
+		// Initializing the datum.
+		d, err := statistics.TopNMetaValToDatum(cur.item.Encoded, hists[0].Tp.GetType(), isIndex, loc)
+		if err != nil {
+			return err
+		}
+		affectedHist = affectedHist[:0]
+		// The following codes might access the NextClear loop twice. Record it here for saving CPU.
+		for histPos, found := cur.affectedTopNs.NextClear(0); found; histPos, found = cur.affectedTopNs.NextClear(histPos + 1) {
+			affectedHist = append(affectedHist, int(histPos))
+		}
+		// Hacking skip.
+		if uint32(len(finalTopNs)) >= n {
```

Comment:
I am not sure which one we prefer. `len(finalTopNs)` vs. `finalTopNs.Len()`. But I guess it doesn't matter.